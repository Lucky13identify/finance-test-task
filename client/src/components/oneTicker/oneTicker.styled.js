import styled from 'styled-components';

const flexSettings = `
  display: flex;
  align-items: center;
  justify-content: center;
`;

export const Container = styled.ul`
  display: flex;

  width: 1000px;
  gap: 10px;

  background-color: #d3d3d3;
  border-radius: 10px;
`;

export const Ticker = styled.p`
  ${flexSettings}

  padding: 3px;
  width: 70px;
  height: 20px;
  border-radius: 3px;

  color: white;
  background-color: ${props => {
    if (props.ticker === 'AAPL') {
      return '#2E8B57';
    } else if (props.ticker === 'GOOGL') {
      return '#DC143C';
    } else if (props.ticker === 'MSFT') {
      return '#808000';
    } else if (props.ticker === 'AMZN') {
      return '#008B8B';
    } else if (props.ticker === 'FB') {
      return '#FF6347';
    } else if (props.ticker === 'TSLA') {
      return '#800080';
    }
  }};
`;

export const Value = styled.li`
  ${flexSettings}

  width: 90px;
  padding: 15px;
`;

export const TextAndTime = styled.li`
  ${flexSettings}

  width: 100px;

  font-style: italic;
`;

export const ProfitOrLoss = styled.p`
  ${flexSettings}

  width: 100%;
  padding: 3px;
  border-radius: 4px;

  font-weight: bold;

  color: ${props => (props.result === 'true' ? 'red' : 'green')};
  background-color: ${props =>
    props.result === 'true' ? 'rgb(252,230,230)' : 'rgb(220,244,220)'};
`;

export const Button = styled.button`
  padding: 10px;

  font-family: inherit;
  font-weight: 700;
  font-size: 16px;

  cursor: pointer;
  border: none;
  border-radius: 4px;

  color: white;
  background-color: #0000cd;
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.15);
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1);

  &:focus,
  &:hover {
    background-color: #00008b;
  }
`;
