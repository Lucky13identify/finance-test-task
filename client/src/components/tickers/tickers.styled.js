import styled from 'styled-components';

export const Container = styled.ul`
  display: flex;
  flex-direction: column;
  gap: 5px;
  margin-top: 20px;
  margin-left: 20px;
`;

export const Head = styled.h1`
  margin-left: 20px;
`;
