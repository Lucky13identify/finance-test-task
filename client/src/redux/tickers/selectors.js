export const selectTickers = state => state.tickers;
export const selectdeletedTickers = state => state.tickers.deletedTickers;
