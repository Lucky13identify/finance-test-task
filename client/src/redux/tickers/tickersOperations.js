import { createAsyncThunk } from '@reduxjs/toolkit';
import io from 'socket.io-client';

const socket = io('http://localhost:4000');

export const fetchTickers = createAsyncThunk('tickers', async () => {
  try {
    socket.emit('start');

    return new Promise(resolve => {
      socket.on('ticker', quotes => {
        resolve(quotes);
      });
    });
  } catch (error) {
    throw error;
  }
});
