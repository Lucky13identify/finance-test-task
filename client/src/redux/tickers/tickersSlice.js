import { createSlice } from '@reduxjs/toolkit';
import { fetchTickers } from './tickersOperations';

const initialState = {
  tickers: [],
  deletedTickers: [],
};

const tickersSlice = createSlice({
  name: 'tickers',
  initialState,
  reducers: {
    removeTicker: (state, action) => {
      state.deletedTickers.push(action.payload);
      state.tickers = state.tickers.filter(
        ticker => ticker.ticker !== action.payload
      );
    },
  },
  extraReducers: builder => {
    builder.addCase(fetchTickers.fulfilled, (state, action) => {
      state.tickers = action.payload.filter(
        ticker => ticker.ticker !== state.deletedTickers
      );
    });
  },
});

export const { removeTicker } = tickersSlice.actions;
export const tickersReducer = tickersSlice.reducer;
