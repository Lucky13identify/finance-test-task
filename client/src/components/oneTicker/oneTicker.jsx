import { useState, useEffect } from 'react';
import { useDispatch } from 'react-redux';
import Notiflix from 'notiflix';
import { CreateCSSTransition } from '../../utils/cssTransition';
import { thousandsSeparator } from '../../utils/formatNumberWithSpaces';
import { replaceDotsWithCommas } from '../../utils/replaceDots';
import { removeTicker } from '../../redux/tickers/tickersSlice';
import { formatTradeDate } from '../../utils/formattedDate';
import {
  Container,
  Value,
  TextAndTime,
  Ticker,
  ProfitOrLoss,
  Button,
} from './oneTicker.styled';

export function OneTicker({ ticker }) {
  const dispatch = useDispatch();
  const [isVisible, setIsVisible] = useState(false);

  const { change_percent, dividend, change, exchange, last_trade_time, price } =
    ticker;

  useEffect(() => {
    setIsVisible(true);
  }, [change]);

  const deleteTicker = () => {
    dispatch(removeTicker(ticker.ticker));
    Notiflix.Notify.info(
      `You have removed the ${ticker.ticker} ticker. To bring it back, please refresh the page.`
    );
  };

  return (
    <>
      <li className="one-ticker">
        <Container>
          <Value>
            <Ticker ticker={ticker.ticker}>{ticker.ticker}</Ticker>
          </Value>
          <Value>
            {CreateCSSTransition(
              isVisible,
              <ProfitOrLoss result={change.includes('-').toString()}>
                {replaceDotsWithCommas(change)} $
              </ProfitOrLoss>
            )}
          </Value>
          <Value>
            {CreateCSSTransition(
              isVisible,
              <ProfitOrLoss result={change.includes('-').toString()}>
                {change.includes('-')
                  ? `↓ ${Math.floor(change_percent * 100)}`
                  : `↑ ${Math.floor(change_percent * 100)}`}
                %
              </ProfitOrLoss>
            )}
          </Value>
          <Value>
            {CreateCSSTransition(
              isVisible,
              <ProfitOrLoss result={change.includes('-').toString()}>
                {change.includes('-')
                  ? `↓ ${replaceDotsWithCommas(ticker.yield)}`
                  : `↑ ${replaceDotsWithCommas(ticker.yield)}`}
              </ProfitOrLoss>
            )}
          </Value>
          <Value>
            {CreateCSSTransition(
              isVisible,
              <p>{replaceDotsWithCommas((dividend * 10).toFixed(2))}</p>
            )}
          </Value>
          <Value>
            {CreateCSSTransition(
              isVisible,
              <p>{thousandsSeparator(price)} $</p>
            )}
          </Value>
          <TextAndTime>
            <p>{exchange}</p>
          </TextAndTime>
          <TextAndTime>
            <p>{formatTradeDate(last_trade_time)} </p>
          </TextAndTime>
          <Value>
            <Button onClick={deleteTicker}>Delete</Button>
          </Value>
        </Container>
      </li>
    </>
  );
}
