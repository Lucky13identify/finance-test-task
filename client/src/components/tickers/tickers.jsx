import { useSelector } from 'react-redux';
import { OneTicker } from '../oneTicker/oneTicker';
import {
  selectTickers,
  selectdeletedTickers,
} from '../../redux/tickers/selectors';
import { Container, Head } from './tickers.styled';

export function TickersList() {
  const tickersStore = useSelector(selectTickers);
  const deletedTickersStore = useSelector(selectdeletedTickers);

  const filteredTickers = tickersStore.tickers.filter(
    ticker => !deletedTickersStore.includes(ticker.ticker)
  );

  if (!tickersStore || !Array.isArray(tickersStore.tickers)) {
    return null;
  }

  return (
    <>
      <Head>Tickers:</Head>
      <Container>
        {filteredTickers.map(ticker => (
          <OneTicker key={ticker.id} ticker={ticker} />
        ))}
      </Container>
    </>
  );
}
