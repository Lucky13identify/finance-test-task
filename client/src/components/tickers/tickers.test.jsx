import { render, screen } from '@testing-library/react';
import { TickersList } from './tickers';
import { data } from '../oneTicker/oneTicker.test';
import { Provider } from 'react-redux';
import { store } from '../../redux/store';

describe('Tickers Component', () => {
  it('should render container of tickers', () => {
    render(
      <Provider store={store}>
        <TickersList key={data.id} ticker={data} />
      </Provider>
    );

    expect(screen.getByText('Tickers:')).toBeInTheDocument();
  });
  it('snapshot', () => {
    const list = render(
      <Provider store={store}>
        <TickersList key={data.id} ticker={data} />
      </Provider>
    );

    expect(list).toMatchSnapshot();
  });
});
