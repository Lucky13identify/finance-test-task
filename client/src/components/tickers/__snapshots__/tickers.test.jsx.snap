// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OneTicker Component snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div
      id="NotiflixNotifyWrap"
      style="width: 280px; z-index: 4001; opacity: 1; right: 10px; top: 10px;"
    >
      <div
        class="notiflix-notify notiflix-notify-info nx-with-animation nx-with-icon nx-fade   "
        id="NotiflixNotify-1"
        style="font-size: 13px; color: rgb(255, 255, 255); background: rgb(38, 192, 211); border-radius: 5px; pointer-events: all; font-family: \\"Quicksand\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, \\"Helvetica Neue\\", Arial, \\"Noto Sans\\", sans-serif; animation-duration: 400ms;"
      >
        <svg
          class="nx-message-icon"
          height="40"
          viewBox="0 0 40 40"
          width="40"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <path
              d="M20 0c11.03,0 20,8.97 20,20 0,11.03 -8.97,20 -20,20 -11.03,0 -20,-8.97 -20,-20 0,-11.03 8.97,-20 20,-20zm0 37.98c9.92,0 17.98,-8.06 17.98,-17.98 0,-9.92 -8.06,-17.98 -17.98,-17.98 -9.92,0 -17.98,8.06 -17.98,17.98 0,9.92 8.06,17.98 17.98,17.98zm-0.99 -23.3c0,-0.54 0.44,-0.98 0.99,-0.98 0.55,0 0.99,0.44 0.99,0.98l0 15.86c0,0.55 -0.44,0.99 -0.99,0.99 -0.55,0 -0.99,-0.44 -0.99,-0.99l0 -15.86zm0 -5.22c0,-0.55 0.44,-0.99 0.99,-0.99 0.55,0 0.99,0.44 0.99,0.99l0 1.09c0,0.54 -0.44,0.99 -0.99,0.99 -0.55,0 -0.99,-0.45 -0.99,-0.99l0 -1.09z"
              fill="rgba(0,0,0,0.2)"
            />
          </g>
        </svg>
        <span
          class="nx-message nx-with-icon"
        >
          You have removed the AAPL ticker. To bring it back, please refresh the page.
        </span>
      </div>
    </div>
    <div>
      <li
        class="one-ticker"
      >
        <ul
          class="sc-bdnyFh bItzgU"
        >
          <li
            class="sc-dlniIP gcShLv"
          >
            <p
              class="sc-gtsqUy jxiwjj"
              ticker="AAPL"
            >
              AAPL
            </p>
          </li>
          <li
            class="sc-dlniIP gcShLv"
          >
            <div
              style="transition: opacity 500ms ease-in-out; opacity: 1;"
            >
              <p
                class="sc-eCAqax bwkcXI"
                result=""
              >
                76,65$
              </p>
            </div>
          </li>
          <li
            class="sc-dlniIP gcShLv"
          >
            <div
              style="transition: opacity 500ms ease-in-out; opacity: 1;"
            >
              <p
                class="sc-eCAqax bwkcXI"
                result=""
              >
                ↑ 6
                %
              </p>
            </div>
          </li>
          <li
            class="sc-dlniIP gcShLv"
          >
            <div
              style="transition: opacity 500ms ease-in-out; opacity: 1;"
            >
              <p
                class="sc-eCAqax bwkcXI"
                result=""
              >
                ↑ 0,91
              </p>
            </div>
          </li>
          <li
            class="sc-dlniIP gcShLv"
          >
            <div
              style="transition: opacity 500ms ease-in-out; opacity: 1;"
            >
              <p>
                2,30
              </p>
            </div>
          </li>
          <li
            class="sc-dlniIP gcShLv"
          >
            <div
              style="transition: opacity 500ms ease-in-out; opacity: 1;"
            >
              <p>
                281 550
                 $
              </p>
            </div>
          </li>
          <li
            class="sc-hKFymg eQwxfH"
          >
            <p>
              NASDAQ
            </p>
          </li>
          <li
            class="sc-hKFymg eQwxfH"
          >
            <p>
              10.27.2023
               
            </p>
          </li>
          <li
            class="sc-dlniIP gcShLv"
          >
            <button
              class="sc-jSFipO dYlyAM"
              type="button"
            >
              Delete
            </button>
          </li>
        </ul>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="one-ticker"
    >
      <ul
        class="sc-bdnyFh bItzgU"
      >
        <li
          class="sc-dlniIP gcShLv"
        >
          <p
            class="sc-gtsqUy jxiwjj"
            ticker="AAPL"
          >
            AAPL
          </p>
        </li>
        <li
          class="sc-dlniIP gcShLv"
        >
          <div
            style="transition: opacity 500ms ease-in-out; opacity: 1;"
          >
            <p
              class="sc-eCAqax bwkcXI"
              result=""
            >
              76,65$
            </p>
          </div>
        </li>
        <li
          class="sc-dlniIP gcShLv"
        >
          <div
            style="transition: opacity 500ms ease-in-out; opacity: 1;"
          >
            <p
              class="sc-eCAqax bwkcXI"
              result=""
            >
              ↑ 6
              %
            </p>
          </div>
        </li>
        <li
          class="sc-dlniIP gcShLv"
        >
          <div
            style="transition: opacity 500ms ease-in-out; opacity: 1;"
          >
            <p
              class="sc-eCAqax bwkcXI"
              result=""
            >
              ↑ 0,91
            </p>
          </div>
        </li>
        <li
          class="sc-dlniIP gcShLv"
        >
          <div
            style="transition: opacity 500ms ease-in-out; opacity: 1;"
          >
            <p>
              2,30
            </p>
          </div>
        </li>
        <li
          class="sc-dlniIP gcShLv"
        >
          <div
            style="transition: opacity 500ms ease-in-out; opacity: 1;"
          >
            <p>
              281 550
               $
            </p>
          </div>
        </li>
        <li
          class="sc-hKFymg eQwxfH"
        >
          <p>
            NASDAQ
          </p>
        </li>
        <li
          class="sc-hKFymg eQwxfH"
        >
          <p>
            10.27.2023
             
          </p>
        </li>
        <li
          class="sc-dlniIP gcShLv"
        >
          <button
            class="sc-jSFipO dYlyAM"
            type="button"
          >
            Delete
          </button>
        </li>
      </ul>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Tickers Component snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div
      id="NotiflixNotifyWrap"
      style="width: 280px; z-index: 4001; opacity: 1; right: 10px; top: 10px;"
    >
      <div
        class="notiflix-notify notiflix-notify-info nx-with-animation nx-with-icon nx-fade   "
        id="NotiflixNotify-1"
        style="font-size: 13px; color: rgb(255, 255, 255); background: rgb(38, 192, 211); border-radius: 5px; pointer-events: all; font-family: \\"Quicksand\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, \\"Helvetica Neue\\", Arial, \\"Noto Sans\\", sans-serif; animation-duration: 400ms;"
      >
        <svg
          class="nx-message-icon"
          height="40"
          viewBox="0 0 40 40"
          width="40"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <path
              d="M20 0c11.03,0 20,8.97 20,20 0,11.03 -8.97,20 -20,20 -11.03,0 -20,-8.97 -20,-20 0,-11.03 8.97,-20 20,-20zm0 37.98c9.92,0 17.98,-8.06 17.98,-17.98 0,-9.92 -8.06,-17.98 -17.98,-17.98 -9.92,0 -17.98,8.06 -17.98,17.98 0,9.92 8.06,17.98 17.98,17.98zm-0.99 -23.3c0,-0.54 0.44,-0.98 0.99,-0.98 0.55,0 0.99,0.44 0.99,0.98l0 15.86c0,0.55 -0.44,0.99 -0.99,0.99 -0.55,0 -0.99,-0.44 -0.99,-0.99l0 -15.86zm0 -5.22c0,-0.55 0.44,-0.99 0.99,-0.99 0.55,0 0.99,0.44 0.99,0.99l0 1.09c0,0.54 -0.44,0.99 -0.99,0.99 -0.55,0 -0.99,-0.45 -0.99,-0.99l0 -1.09z"
              fill="rgba(0,0,0,0.2)"
            />
          </g>
        </svg>
        <span
          class="nx-message nx-with-icon"
        >
          You have removed the AAPL ticker. To bring it back, please refresh the page.
        </span>
      </div>
    </div>
    <div>
      <h1
        class="sc-iCoHzw kPSEfO"
      >
        Tickers:
      </h1>
      <ul
        class="sc-gKAaef jRiajQ"
      />
    </div>
  </body>,
  "container": <div>
    <h1
      class="sc-iCoHzw kPSEfO"
    >
      Tickers:
    </h1>
    <ul
      class="sc-gKAaef jRiajQ"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
